File: Medium/UniquePaths/UniquePaths.cpp
// https://leetcode.com/problems/unique-paths/description/

class Solution {
public:
    int uniquePaths(int m, int n) {
        int totalMoves = m+n-2;
        long long result = 1;

        int smallerMoves = min(m,n)-1;

        for(int i = 1; i<=smallerMoves; i++){
            result*=(totalMoves+1-i);
            result/=i;
        }
        return result;
    }

    // int uniquePaths(int m, int n) {
    //     int result = 0;
    //     int currM = 0;
    //     int currN = 0;
    //     traverse(currM, currN, m, n, result);
    //     return result;
    // }

    // void traverse(int currM, int currN, int m, int n, int &result){
    //     if(currM < m-1){
    //         traverse(currM+1, currN, m, n, result);
    //     }
    //     if(currN < n-1){
    //         traverse(currM, currN+1, m, n, result);
    //     }

    //     if(currM==m-1 && currN==n-1){
    //         result+=1;
    //     }
    // }
};

File: Medium/UniquePaths/UniquePaths.py
# https://leetcode.com/problems/unique-paths/

class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        dp = [[0 for _ in range(n+1)] for _ in range(m+1)]
        dp[m-1][n-1] = 1

        for i in range(m-1,-1,-1):
            for j in range(n-1,-1,-1):
                if i!=m-1 or j!=n-1:
                    dp[i][j] = dp[i+1][j] + dp[i][j+1]

        return dp[0][0]
